File: packages/dart/lib/Seguranca/Ataques/AtaqueCorrelacaoPosicoes.dart
import 'dart:math';
import 'dart:typed_data';

import 'package:criptografia/Seguranca/AlvoCriptografico.dart';
import 'package:criptografia/Seguranca/Ataque.dart';
import 'package:criptografia/Seguranca/ResultadoAtaque.dart';
import 'package:criptografia/Seguranca/Util.dart';

/// Verifica se posições DIFERENTES dentro do mesmo bloco de 32 bytes saem
/// correlacionadas (ex: posição 0 acompanha posição 16). Uma correlação entre
/// posições de saída é exatamente o tipo de estrutura que o antigo bug da
/// "metade igual" produzia - e que a autocorrelação temporal pode não pegar.
class AtaqueCorrelacaoPosicoes extends Ataque {
  final int _amostras;
  final int _tamanhoBloco;
  final double _limiteCorrelacao;

  AtaqueCorrelacaoPosicoes({
    int amostras = 3000,
    int tamanhoBloco = 32,
    double limiteCorrelacao = 0.15,
  }) : _amostras = amostras,
       _tamanhoBloco = tamanhoBloco,
       _limiteCorrelacao = limiteCorrelacao;

  @override
  String nome() => 'Correlação entre posições do bloco';

  @override
  ResultadoAtaque executar(AlvoCriptografico alvo) {
    final chave = alvo.chaveDeTeste();

    final blocos = <Uint8List>[];
    for (var i = 0; i < _amostras; i++) {
      blocos.add(
        alvo.gerarKeystreamBruto(
          chave,
          randomBytes(alvo.tamanhoIv()),
          'enc',
          _tamanhoBloco,
        ),
      );
    }

    final suspeitos = <String>[];
    var pior = 0.0;

    for (var a = 0; a < _tamanhoBloco; a++) {
      for (var b = a + 1; b < _tamanhoBloco; b++) {
        final corr = _pearson(blocos, a, b);
        if (corr.abs() > pior.abs()) {
          pior = corr;
        }
        if (corr.abs() > _limiteCorrelacao) {
          suspeitos.add('posições $a-$b (r=${corr.toStringAsFixed(3)})');
        }
      }
    }

    if (suspeitos.isNotEmpty) {
      return ResultadoAtaque(
        nome(),
        true,
        Severidade.alta,
        '${suspeitos.length} par(es) correlacionado(s): '
            '${suspeitos.take(10).join('; ')}',
        {'suspeitos': suspeitos},
      );
    }

    return ResultadoAtaque(
      nome(),
      false,
      Severidade.info,
      'Nenhum par de posições correlacionado acima de '
          '${_limiteCorrelacao.toStringAsFixed(2)} '
          '(pior |r|=${pior.abs().toStringAsFixed(3)})',
    );
  }

  double _pearson(List<Uint8List> blocos, int a, int b) {
    final n = blocos.length;
    var ma = 0.0;
    var mb = 0.0;
    for (final d in blocos) {
      ma += d[a];
      mb += d[b];
    }
    ma /= n;
    mb /= n;

    var num = 0.0;
    var da = 0.0;
    var db = 0.0;
    for (final d in blocos) {
      final xa = d[a] - ma;
      final xb = d[b] - mb;
      num += xa * xb;
      da += xa * xa;
      db += xb * xb;
    }

    return da > 0 && db > 0 ? num / sqrt(da * db) : 0.0;
  }
}

File: packages/rust/src/seguranca/ataques/ataque_correlacao_posicoes.rs
//! Verifica se posições DIFERENTES dentro do mesmo bloco de 32 bytes saem
//! correlacionadas (ex: posição 0 acompanha posição 16). Uma correlação entre
//! posições de saída é exatamente o tipo de estrutura que o antigo bug da
//! "metade igual" produzia - e que a autocorrelação temporal pode não pegar.

use crate::seguranca::alvo_criptografico::AlvoCriptografico;
use crate::seguranca::ataque::{Ataque, ErroAtaque};
use crate::seguranca::resultado_ataque::{ResultadoAtaque, Severidade};
use crate::seguranca::util::random_bytes;

pub struct AtaqueCorrelacaoPosicoes {
    pub amostras: usize,
    pub tamanho_bloco: usize,
    pub limite_correlacao: f64,
}

impl AtaqueCorrelacaoPosicoes {
    pub fn new() -> Self {
        AtaqueCorrelacaoPosicoes {
            amostras: 3000,
            tamanho_bloco: 32,
            limite_correlacao: 0.15,
        }
    }
}

impl Default for AtaqueCorrelacaoPosicoes {
    fn default() -> Self {
        Self::new()
    }
}

impl AtaqueCorrelacaoPosicoes {
    fn pearson(blocos: &[Vec<u8>], a: usize, b: usize) -> f64 {
        let n = blocos.len();
        let mut ma = 0.0f64;
        let mut mb = 0.0f64;
        for d in blocos {
            ma += d[a] as f64;
            mb += d[b] as f64;
        }
        ma /= n as f64;
        mb /= n as f64;

        let mut num = 0.0f64;
        let mut da = 0.0f64;
        let mut db = 0.0f64;
        for d in blocos {
            let xa = d[a] as f64 - ma;
            let xb = d[b] as f64 - mb;
            num += xa * xb;
            da += xa * xa;
            db += xb * xb;
        }

        if da > 0.0 && db > 0.0 {
            num / (da * db).sqrt()
        } else {
            0.0
        }
    }
}

impl Ataque for AtaqueCorrelacaoPosicoes {
    fn nome(&self) -> String {
        "Correlação entre posições do bloco".to_string()
    }

    fn executar(&self, alvo: &dyn AlvoCriptografico) -> Result<ResultadoAtaque, ErroAtaque> {
        let chave = alvo.chave_de_teste();

        let mut blocos: Vec<Vec<u8>> = Vec::with_capacity(self.amostras);
        for _ in 0..self.amostras {
            blocos.push(alvo.gerar_keystream_bruto(
                chave.as_bytes(),
                &random_bytes(alvo.tamanho_iv()),
                "enc",
                self.tamanho_bloco,
            ));
        }

        let mut suspeitos: Vec<String> = Vec::new();
        let mut pior = 0.0f64;

        for a in 0..self.tamanho_bloco {
            for b in (a + 1)..self.tamanho_bloco {
                let corr = Self::pearson(&blocos, a, b);
                if corr.abs() > pior.abs() {
                    pior = corr;
                }
                if corr.abs() > self.limite_correlacao {
                    suspeitos.push(format!("posições {}-{} (r={:.3})", a, b, corr));
                }
            }
        }

        if !suspeitos.is_empty() {
            let amostra: Vec<String> = suspeitos.iter().take(10).cloned().collect();
            return Ok(ResultadoAtaque::new(
                &self.nome(),
                true,
                Severidade::Alta,
                &format!(
                    "{} par(es) correlacionado(s): {}",
                    suspeitos.len(),
                    amostra.join("; ")
                ),
            ));
        }

        Ok(ResultadoAtaque::new(
            &self.nome(),
            false,
            Severidade::Info,
            &format!(
                "Nenhum par de posições correlacionado acima de {:.2} (pior |r|={:.3})",
                self.limite_correlacao,
                pior.abs()
            ),
        ))
    }
}
